docs/event-processing/idempotent-reader.md

```diff
@@ -0,0 +1,52 @@
+
+On the reading side, in [Event Processors](../event-processing/event-processor.md) and [Event Sinks](../event-sink/event-sink.md), an Idempotent Reader can be configured to read only committed transactions. This prevents Events within incomplete transactions from being read, providing the reader isolation from operational writer failures. Keep in mind that idempotency means that the reader's business logic must be able to process the same consumed Event multiple times, without generating side-effects or otherwise incorrectly updating its internal state. 
+
+Duplicates caused by incorrect application logic are best resolved by fixing the application's logic. In cases where this is not possible, such as when event generation is triggered by external processes, the next best option is to tag duplicate events with tracking IDs. A tracking ID should be a field that is unique to the logical event, such as an event key or request ID. The consumer can then read the tracking ID, cross-reference it against an internal state store of IDs it has already processed, and discard the event if necessary.
+
+
+## Implementation
+To handle operational failures, you can [enable EOS in your Kafka Streams application](https://www.confluent.io/blog/enabling-exactly-once-kafka-streams/). A Streams application using EOS will atomically update its consumer offsets, state store changelog topics, reparition topics, and output topics within a single transaction.
+
+In the streaming database [ksqlDB](https://ksqldb.io), you can enable exactly-once stream processing to execute a read-process-write operation exactly one time. This can be done by configuring the processing guarantee with the following setting:
+
+```
+processing.guarantee="exactly_once"
+``` 
+
+To handle incorrect application logic, again, first try to eliminate the source of duplication from the code. If that is not an option, you can assign a tracking ID to each Event based off of the contents of the Event, enabling consumers to detect duplicates for themselves. This requires that each consumer application maintain an internal state store for tracking the Events' unique IDs, and this store will vary in size depending on the Event count and the period for which the consumer must guard against duplicates. This option requires both additional disk usage and processing power for inserting and validating Events.
+
+For a subset of business cases, it may also be possible to design the consumer processing logic to be idempotent. For example, in a simple ETL where a database is the [Event Sink](../event-sink/event-sink.md), the duplicate Events can be deduplicated by the database during an `upsert` on the Event ID as primary key. Idempotent business logic also enables your application to read from [Event Streams](../event-stream/event-stream.md) that aren't strictly free of duplicates, or from historic streams that may not have had EOS available when they were created.
```

Comment:
```suggestion
For a subset of business cases, it may also be possible to design the consumer processing logic to be idempotent. For example, in a simple ETL where a database is the [Event Sink](../event-sink/event-sink.md), the duplicate events can be deduplicated by the database during an `upsert` on the event ID as primary key. Idempotent business logic also enables your application to read from [Event Streams](../event-stream/event-stream.md) that aren't strictly free of duplicates, or from historic streams that may not have had EOS available when they were created.
```